component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/validators/DPoPValidator.java

```diff
@@ -0,0 +1,276 @@
+import org.wso2.carbon.identity.oauth2.dto.OAuth2TokenValidationResponseDTO;
+import org.wso2.carbon.identity.oauth2.model.HttpRequestHeader;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+import org.wso2.carbon.identity.oauth2.token.bindings.TokenBinding;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+
+import java.security.interfaces.ECPublicKey;
+import java.security.interfaces.RSAPublicKey;
+import java.sql.Timestamp;
+import java.text.ParseException;
+import java.util.Date;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+
+public class DPoPValidator {
+
+    static final Log log = LogFactory.getLog(DPoPValidator.class);
+
+    public static AuthenticationResult validateDPoPHeaderandToken(AuthenticationResult authenticationResult,
+                                                                  OAuth2TokenValidationResponseDTO responseDTO,
+                                                                  AuthenticationRequest authenticationRequest) {
+
+        if (!responseDTO.isValid()) {
+            log.debug(responseDTO.getErrorMsg());
+            return authenticationResult;
+        }
+        DPoPBasedTokenBinder dPoPBasedTokenBinder = new DPoPBasedTokenBinder();
+        HttpServletRequest request = authenticationRequest.getRequest();
+        if (dPoPBasedTokenBinder.isValidTokenBinding(request, responseDTO.getTokenBinding())) {
+            authenticationResult.setAuthenticationStatus(AuthenticationStatus.SUCCESS);
+        }
+        return authenticationResult;
+    }
+
+    public static boolean isValidDPoP(String dPoPProof, OAuth2AccessTokenReqDTO tokenReqDTO,
+                                      OAuthTokenReqMessageContext tokReqMsgCtx)
+            throws IdentityOAuth2Exception {
+
+        try {
+            SignedJWT signedJwt = SignedJWT.parse(dPoPProof);
+            JWSHeader header = signedJwt.getHeader();
+            validateDPoPHeader(header);
+            validateDPoPPayload(tokenReqDTO, signedJwt.getJWTClaimsSet());
+            return validateSignature(signedJwt, tokReqMsgCtx);
+        } catch (ParseException | JOSEException e) {
+            throw new IdentityOAuth2ClientException(DPoPConstants.INVALID_DPOP_PROOF, DPoPConstants.INVALID_DPOP_ERROR);
+        }
+    }
+
+    private static boolean validateSignature(SignedJWT signedJwt, OAuthTokenReqMessageContext tokReqMsgCtx)
+            throws ParseException, JOSEException, IdentityOAuth2Exception {
+
+        JWK jwk = JWK.parse(signedJwt.getHeader().getJWK().toString());
+        TokenBinding tokenBinding = new TokenBinding();
+        tokenBinding.setBindingType(DPoPConstants.DPOP_TOKEN_TYPE);
+        boolean isValid = false;
+
+        // Using the EC algorithm.
+        if (DPoPConstants.ECDSA_ENCRYPTION.equalsIgnoreCase(jwk.getKeyType().toString())) {
+            ECKey ecKey = (ECKey) jwk;
```

Comment:
have this logic in separate private methods for better readabilty

Reply:
fixed